Global Styles: Try per block custom CSS

carolinan opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Adds the option to add custom CSS per block via the Styles sidebar in the Site Editor and via theme.json.
Continuation on https://github.com/WordPress/gutenberg/issues/30142, https://github.com/WordPress/gutenberg/issues/44412

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
To allow more customizations.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
This PR builds on the previous changes that adds the option to add custom CSS to the root via the Styles sidebar.

- Updates the global styles **context menu** to display the button that opens the Custom CSS panel for each block.
- Updates the **CustomCSSControl** to use the block specific CSS (styles.blocks.blockname.css) if a block name is defined.
-  Updates the screen, **ScreenCSS** to pass the block name to the CustomCSSCControl and to display a slightly different panel description, containing the block name (I added the block name for clarity and ease of use ).
- Outputs the per block CSS in the editors and front.

## Testing Instructions

1. Apply the PR
2. Activate the test theme, "empty theme"
3. In theme.json, add CSS to `styles.blocks.blockname.css`,
for example to the paragraph block:

```
"styles": {
		"blocks": {
			"core/paragraph": {
				"css": "color: orange;"
			}
		}
	}
```
4. View the theme and confirm that the CSS you added in theme.json is applied in the site editor, block editor, and on the front.
5. Open the Site Editor. Select "Edit" in the new navigation sidebar. Open the Styles sidebar, select Blocks.
Open the Styles sidebar, select Blocks.
6. Select the block that you added CSS to in the previous step.
7. Below the typography, colors and layout settings you should find a new panel with the text "Add your own CSS to customize the block appearance."
8. Select the button with the text "Additional block CSS" to open the CSS panel.
9. Confirm that there is a description at the top of the panel, after the back button and Heading.
Confirm that the description contains the name of the current block:
"Add your own CSS to customize the appearance of the Paragraph block."
10. Below that is an h2 heading that says "ADDITIONAL CSS" followed by 
a text area for inputting CSS.
11. Confirm that the text area shows the CSS that you added to theme.json.
12. Make a change to the existing CSS and make a new addition.
13. Confirm that the changes are applied to the blocks in the site editor.
14. Confirm that there is a help text below the text area with some instructions.
15. Next, confirm that there is an additional panel below the help text.
- The panel should be closed by default.
- The panel has an H2 heading with a button with the text "Original Theme Custom CSS"
Activate the button to open the panel.
The panel should contain a `<pre>` element with the CSS that you added to theme.json.
16. Save the style changes.
17. Go back one step to the previous panel, the panel for the block.
18. Confirm that the block preview has been updated with your styles.
19. View the front of the website and confirm that your styles are applied.


### Testing Instructions for Keyboard
The Styles sidebar in general has accessibility problems where the heading order is incorrect and where information
is not available to screen reader users.
These problems affect the sidebar as a whole and needs to be carefully tested and fixed in a separate PR.

### Testing permissions
1. Disable unfiltered HTML by adding the following constant to wp-config.php - define( 'DISALLOW_UNFILTERED_HTML', true );
2. Refresh the Site Editor, open the Styles sidebar, select Blocks. Select a block, confirm that this panel does not have a custom CSS button.

1. Apply the PR
2. Activate the test theme, "empty theme"
3. In theme.json, add CSS to `styles.blocks.blockname.css`,
for example to the paragraph block:

```
"styles": {
		"blocks": {
			"core/paragraph": {
				"css": "color: orange;"
			}
		}
	}
```
4. From the WordPress admin area, navigate to the Site Editor under Appearance, Editor.
5. Once in the Site Editor, activate the button with the aria-label "Open the editor".
6. Navigate backwards into the "Editor top toolbar". Navigate backwards past the Options and Navigation buttons, and activate the "Styles" button.
7 - Note: The styles sidebar is now opened, but it is unclear to me how to best navigate to it because activating the button does not move focus to the sidebar content.
Instead, I needed to tab forward past the navigation and option buttons, and past the editor content.
The next focusable item is then a button labeled "open Style book". This button is inside the styles panel, but it is not the button we want to access.
Instead, press the tab key to tab forwards past More styles, close, browse styles, typography, colors, layout, until you arrive at the button "Block Styles". Activate this button.
8. _Even though it is not announced that the block styles panel is opened,_ focus should now be inside the correct sidebar panel. Use tab to navigate through a list of buttons labeled with the block name. Select the button for the block that you edited in theme.json.
9. Navigate past the block options for typography, color and layout until you reach the button with the aria-label "Additional block CSS". Activate the button.
10. tab forwards to reach a text area. When the text area is in focus, a help text with instructions is also read out.
11. Confirm that there is CSS in the text area that matches what you added to theme.json.
12. Edit the CSS and make your own addition.
13. Save.
14. Reload/refresh the page and navigate back to the text area. Confirm that the updates you made have been saved.

## Screenshots or screencast <!-- if applicable -->

<img width="279" alt="The button that opens the custom CSS panel, in the Styles > blocks panel" src="https://user-images.githubusercontent.com/7422055/207802193-a03dcae0-ab07-4fec-b585-3bb336c18701.png">


<img width="279" alt="The per block custom CSS panel in the site editor" src="https://user-images.githubusercontent.com/7422055/207801971-8b3419bd-aea5-478c-a9bc-7d9e19027656.png">

carolinan committed:
WIP, try adding per block custom css to global styles

carolinan committed:
Update class-wp-theme-json-6-2.php

carolinan committed:
Update use-global-styles-output.js

carolinan committed:
Add the CSS selector to the global styles output in the editor.

carolinan committed:
Update use-global-styles-output.js

aristath committed:
Process CSS nesting

glendaviesnz commented:
🎉  This is looking good! I like the simple approach arrived at of not needing the `:self` style wrapper selector and the `&` nesting. Do you think we should add some help text above the input box about the use of `&` for pseudo and nested selectors?

carolinan commented:
I would like that, but how can we describe that so that users of all varied experiences can understand it?

carolinan commented:
The changes need to be moved from lib/compat/wordpress-6.2/class-wp-theme-json-6-2.php to class-wp-theme-json-gutenberg.php, but I am away for the holiday starting about now and I welcome anyone to pick it up.

glendaviesnz commented:
Just adding a note that this will also be impacted by [this issue](https://github.com/WordPress/gutenberg/issues/46651) - so will need to account for a user's unfiltered HTML capability somehow.

aristath commented:
> The changes need to be moved from lib/compat/wordpress-6.2/class-wp-theme-json-6-2.php to class-wp-theme-json-gutenberg.php, but I am away for the holiday starting about now and I welcome anyone to pick it up.

Rebased the PR on trunk. Moved the code to `class-wp-theme-json-gutenberg.php` and tested the implementation, everything is still working as expected 👍

Mamaduka commented:
We can reuse logic from #46815 for the unfiltered HTML capability check.

carolinan commented:
The permission check should be added to context-menu.js, but the new action needs to be merged first.

carolinan committed:
Merge branch 'trunk' into try/per-block-custom-css

carolinan committed:
Add permission check

Check for permission to edit CSS before displaying the panel.
Fix duplication after merge conflict.

carolinan commented on packages/edit-site/src/components/global-styles/context-menu.js:
```diff
@@ -24,48 +41,97 @@ function ContextMenu( { name, parentMenu = '' } ) {
+				{ hasVariationsPanel && (
+					<ScreenVariations name={ name } path={ parentMenu } />
+				) }
+				{ !! parentMenu && !! canEditCSS && (
```
With this condition, the panel with the button that opens the custom CSS panel shows when you navigate to Styles > Blocks > Block but also when the block's style variation panel is opened ( Styles > Blocks > Block > Style Variation ).
You can test this by selecting the image block or any other block that has a custom block style.
How can I prevent the CSS panel from displaying in the style variation panel?

@Mamaduka @tellthemachines

Mamaduka replied:
@carolinan, this is on my to-do list. I will try to get back to you with the answer ASAP.

Mamaduka replied:
```suggestion
				{  isBlocksPanel && canEditCSS && (
```

I think we can use `parentMenu` to determine the current panel. The variations panel [uses a similar check](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/global-styles/variations-panel.js#L19).

```js
const isBlocksPanel =
		parentMenu.includes( 'blocks' ) &&
		! parentMenu.includes( 'variations' );
```

P.S. There's no need for a double exclamation before `canEditCSS` it's already a boolean value.

carolinan committed:
Fix CS issue, spacing

carolinan commented on packages/edit-site/src/components/global-styles/context-menu.js:
```diff
@@ -24,48 +41,97 @@ function ContextMenu( { name, parentMenu = '' } ) {
 	const hasLayoutPanel = hasDimensionsPanel;
 	const hasVariationsPanel = useHasVariationsPanel( name, parentMenu );
 
+	const { canEditCSS } = useSelect( ( select ) => {
```
There is some code duplication here from [screen-root.js](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/global-styles/screen-root.js#L29), I could not figure out how or if I could pass the `canEditCSS` value from the screen root to the next panel via the `<NavigationButtonAsItem>` ?

Mamaduka replied:
@carolinan, probably it's not worth it. We've hooks like `useSelect` to avoid prop drilling.

Mamaduka commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -956,6 +956,27 @@ public function get_settings() {
+	 *
+	 * @return string The processed CSS.
+	 */
+	private function process_nested_css( $css, $selector ) {
+		$processed_css = '';
+
+		// Split CSS nested rules.
+		$parts = explode( '&', $css );
+		foreach ( $parts as $part ) {
+			$processed_css .= ( ! str_contains( $part, '{' ) )
+				? $selector . '{' . $part . '}' // If the part doesn't contain braces, it applies to the root level.
+				: $selector . $part; // Prepend the selector, which effectively replaces the "&" character.
+		}
+		return $processed_css;
+	}
```
It would be good to have unit tests for this new method. We can't test private methods, but we can check the `$theme_json->get_stylesheet()` output. See `test_get_stylesheet_handles_custom_css` for examples.

Would this method be applicable outside of block custom CSS? Maybe we should use a more explicit name for it. What do you think?

aristath replied:
> Would this method be applicable outside of block custom CSS? Maybe we should use a more explicit name for it. What do you think?

I can't imagine where this would be used outside the block custom-css... I'll rename it to something more specific and write some unit tests 👍

aristath replied:
method renamed to `process_blocks_custom_css_nested_styles`, and PHPUnit tests added 👍

Mamaduka commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -1062,7 +1083,19 @@ public function get_stylesheet( $types = array( 'variables', 'styles', 'presets'
+			// Add the global styles block CSS.
+			if ( isset( $this->theme_json['styles']['blocks'] ) ) {
+				foreach ( $this->theme_json['styles']['blocks'] as $name => $node ) {
+					if ( _wp_array_get( $this->theme_json, array( 'styles', 'blocks', $name, 'css' ) ) ) {
+						$selector         = static::$blocks_metadata[ $name ]['selector'];
+						$custom_block_css = _wp_array_get( $this->theme_json, array( 'styles', 'blocks', $name, 'css' ) );
+						$stylesheet      .= $this->process_nested_css( $custom_block_css, $selector );
+					}
```
Suggestion (non-block): There's no need to get the same value twice here.

```php
$custom_block_css = _wp_array_get( $this->theme_json, array( 'styles', 'blocks', $name, 'css' ) );

if ( $custom_block_css ) {
  $selector         = static::$blocks_metadata[ $name ]['selector'];
  $stylesheet      .= $this->process_nested_css( $custom_block_css, $selector );
}
```

carolinan committed:
Update context menu, simplify condition when loading the CSS

aristath committed:
rename method

aristath committed:
make method public

aristath committed:
Add unit tests

Mamaduka commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -956,6 +956,27 @@ public function get_settings() {
+	 *
+	 * @return string The processed CSS.
+	 */
+	public function process_blocks_custom_css_nested_styles( $css, $selector ) {
```
Sorry for nitpicking here. The "nested styles" is the result of processing, right? In that case, would it make sense to call this method `process_blocks_custom_css`?

aristath replied:
makes sense... renamed and committed 👍

glendaviesnz commented:
This is testing well for me, and I also checked without `unfiltered_html` capability, and with `add_action( 'init', 'kses_init_filters' );` set and it worked as expected.

aristath committed:
rename method

Mamaduka commented on packages/edit-site/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -1049,6 +1064,35 @@ export function useGlobalStylesOutput() {
 			},
 		];
 
+		const processCSSNesting = ( css, blockSelector ) => {
+			let processedCSS = '';
+
+			// Split CSS nested rules.
+			const parts = css.split( '&' );
+			parts.forEach( ( part ) => {
+				processedCSS += ! part.includes( '{' )
+					? blockSelector + '{' + part + '}' // If the part doesn't contain braces, it applies to the root level.
+					: blockSelector + part; // Prepend the selector, which effectively replaces the "&" character.
+			} );
+			return processedCSS;
+		};
```
Suggestion (non-blocking): This function can be moved outside the component — no need to recreate it on every re-render.

aristath replied:
Done, tested, and confirmed everything still works fine 👍

Mamaduka commented on packages/edit-site/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -1016,10 +1028,13 @@ export function useGlobalStylesOutput() {
 			getBlockTypes(),
 			getBlockStyles
 		);
+		const blockNames = getBlockNames( getBlockTypes() );
```
Suggestion (non-blocking): This could be "simplified" to something like this:

```js
getBlockTypes().map( ( blockType ) => blockType.name );
```

Then we can directly use the array below to push block stylesheets. No need for `Object.entries( blockNames )`.

aristath replied:
good catch! Done 👍

Mamaduka replied:
It looks like you had a better idea than me in [ccdc5df](https://github.com/WordPress/gutenberg/pull/46571/commits/ccdc5df59f66f39fb6948b34e690c03b3b86b3cf). You can also drop the `.map()` and loop over the block object directly.

Mamaduka commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -956,6 +956,27 @@ public function get_settings() {
+	 *
+	 * @return string The processed CSS.
+	 */
+	public function process_blocks_custom_css( $css, $selector ) {
```
Are we sure we want this method to be public?

aristath replied:
Well... In the beginning, I made it public to make testing the method easier. 
But the more I think about it, the more I like it... The method doesn't change anything in the object itself, so I don't see a reason to make it private, as it's pretty self-isolated. This way it will be possible for 3rd parties to use it in their own implementations if they choose to do something with nested CSS. In fact, we could even make it a static function and nothing would change. Perhaps it would even be preferable? 🤷

Mamaduka replied:
That's the reason I asked. If the method is `public` just for unit tests, then we can make it `protected` and test the expected output using `get_stylesheet()`. We already do this for global custom CSS. 

If you think it could be helpful outside of this particular case, then sure, keeping it public makes sense.

oandregal replied:
> This way it will be possible for 3rd parties to use it in their own implementations if they choose to do something with nested CSS. In fact, we could even make it a static function and nothing would change. Perhaps it would even be preferable? 

Note `WP_Theme_JSON` and `WP_Theme_JsON_Resolver` are private classes, which means we're asking people not to use them. The API should provide public methods for any operation third-parties may need. See related conversation at https://github.com/WordPress/gutenberg/issues/45171

Would you mind making this protected?

Mamaduka reviewed (approved):
Fantastic work, @carolinan, and @aristath!

Let's resolve the merge conflict and get this feature in.

I also left a few suggestions that can also be addressed in a follow-up PR 👍

aristath committed:
Move processCSSNesting function

aristath committed:
simplify code - props @mamaduka

aristath committed:
Merge branch 'trunk' into try/per-block-custom-css

aristath committed:
Simplify the loop some more - props @mamaduka

Mamaduka commented:
@carolinan, @aristath, do you have time to work on core backports for changes in the `WP_Theme_JSON` file? If you don't, just let me know, and happy to do it myself.

aristath commented:
Backporting to Core PR: https://github.com/WordPress/wordpress-develop/pull/3977 for ticket https://core.trac.wordpress.org/ticket/57621

carolinan commented:
### Dev note

When a theme that has support for the Site Editor is activated, users do not have easy access to adding additional CSS via the Customizer.
In WordPress 6.2, there are two new ways to add per-block custom CSS: Via the Styles interface in the Site Editor and theme.json.

To add CSS via the interface, open the Styles sidebar in the Site Editor.
Next, open the blocks panel and select a block. Click on Additional block CSS to open the block CSS panel.

https://user-images.githubusercontent.com/7422055/219643226-d7657d8e-d52d-4c01-93fc-3f1acdbef127.mov

CSS added in this panel applies to all occurrences of the block.
In this panel, you can also see if the active theme adds CSS using theme.json:
<img width="278" alt="The per-block custom CSS panel displays the CSS added in theme.json." src="https://user-images.githubusercontent.com/7422055/219643582-cb0a4645-b6e2-4a78-9ed8-081cbb21dd8e.png">

You can add per-block custom CSS in theme.json as a `string` in `styles.blocks.block.css`.
Basic example:

```
"styles": {
	"blocks": {
		"core/button": {
			"css": "background: purple"
		}
	}
}
```
In addition, the use of & is supported for nested elements & pseudo-selectors, both in the theme.json file and the custom CSS input in global styles.
```
"styles": {
	"blocks": {
		"core/group": {
			"css": "background:var(--wp--preset--color--base); & .cta { background:#fafafa; padding:var(--wp--preset--spacing--40); } & .wp-element-button { background-color:purple; color: white; } & .wp-element-button:hover { opacity: 0.9; }"
		}
	}
}
```

-----
Reviewed by @aristath

mrleemon commented:
@carolinan Does styling via `theme.json` work for non-core custom blocks? I'm trying the `css` prop on a custom block in my custom theme and the styles only appear in the block editor, not the frontend.

carolinan commented:
Hi,
If you believe it is a bug please open a new issue with the full details. For example, is it a classic theme or block theme, is it a dynamic block, does the other theme.json styles work, and so on.
First, test with a different theme and block.

mrleemon commented:
@carolinan I opened a bug report: https://github.com/WordPress/gutenberg/issues/52644
It's happening with core blocks, too.
And block themes are not affected, just classic ones.
